File: android/src/main/kotlin/com/example/health_messenger_ui/HealthMessengerUiPlugin.kt
package com.example.health_messenger_ui

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class HealthMessengerUiPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private var applicationContext: Context? = null
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext

        val messenger = binding.binaryMessenger
        val ch = MethodChannel(messenger, CHANNEL)
        ch.setMethodCallHandler(this)
        methodChannel = ch

        val ev = EventChannel(messenger, EVENT_CHANNEL)
        ev.setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                    synchronized(staticLock) {
                        staticSink = events
                    }
                }

                override fun onCancel(arguments: Any?) {
                    synchronized(staticLock) {
                        staticSink = null
                    }
                }
            },
        )
        eventChannel = ev
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        applicationContext = null
        synchronized(staticLock) {
            staticSink = null
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val ctx = applicationContext
        if (ctx == null) {
            result.error("no_context", "Plugin not attached", null)
            return
        }
        when (call.method) {
            "syncPushConfig" -> {
                @Suppress("UNCHECKED_CAST")
                val args = call.arguments as? Map<*, *> ?: emptyMap<Any, Any>()
                PushNativeStore.syncPushConfig(ctx, args)
                result.success(null)
            }
            "drainAckQueue" -> {
                Thread {
                    val config = PushNativeStore.readConfig(ctx)
                    if (config == null) {
                        Handler(Looper.getMainLooper()).post { result.success(0) }
                        return@Thread
                    }
                    var done = 0
                    val pending = PushNativeStore.dequeueAll(ctx)
                    for ((conversationId, messageId) in pending) {
                        if (NativeDeliveredAck.postDelivered(config, conversationId, messageId)) {
                            done++
                        } else {
                            PushNativeStore.enqueueAck(ctx, conversationId, messageId)
                            break
                        }
                    }
                    Handler(Looper.getMainLooper()).post { result.success(done) }
                }.start()
            }
            else -> result.notImplemented()
        }
    }

    companion object {
        private const val CHANNEL = "health_messenger_ui/push"
        private const val EVENT_CHANNEL = "health_messenger_ui/push_events"

        private val staticLock = Any()
        private var staticSink: EventChannel.EventSink? = null

        fun emitIncomingPush(payload: Map<String, Any>) {
            val sink = synchronized(staticLock) { staticSink } ?: return
            Handler(Looper.getMainLooper()).post {
                sink.success(payload)
            }
        }
    }
}

File: ios/Classes/HealthMessengerUiPlugin.swift
import Flutter
import UIKit

@objcMembers
public class HealthMessengerUiPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    let method = FlutterMethodChannel(name: "health_messenger_ui/push", binaryMessenger: messenger)
    let instance = HealthMessengerUiPlugin()
    method.setMethodCallHandler { call, result in
      instance.handle(call, result: result)
    }
    let events = FlutterEventChannel(name: "health_messenger_ui/push_events", binaryMessenger: messenger)
    events.setStreamHandler(PushEventStreamHandler())
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "syncPushConfig":
      if let args = call.arguments as? [String: Any] {
        PushNativeStore.syncPushConfig(args)
      }
      result(nil)
    case "drainAckQueue":
      DispatchQueue.global(qos: .utility).async {
        guard let config = PushNativeStore.readConfig() else {
          DispatchQueue.main.async { result(0) }
          return
        }
        var done = 0
        let pending = PushNativeStore.dequeueAll()
        for item in pending {
          if NativeDeliveredAck.postDelivered(config: config, conversationId: item.0, messageId: item.1) {
            done += 1
          } else {
            PushNativeStore.enqueueAck(conversationId: item.0, messageId: item.1)
            break
          }
        }
        DispatchQueue.main.async { result(done) }
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Call from `AppDelegate.application(_:didReceiveRemoteNotification:fetchCompletionHandler:)`.
  /// Returns `true` when the payload was recognized as a chat message push.
  @objc public static func handleRemoteNotification(
    userInfo: [AnyHashable: Any],
    completionHandler: @escaping (UIBackgroundFetchResult) -> Void
  ) -> Bool {
    let flat = flattenUserInfo(userInfo)
    guard let config = PushNativeStore.readConfig() else {
      return false
    }
    let typeKey = config.chatTypeDataKey
    let incoming = flat[typeKey] ?? ""
    if incoming != config.chatTypeValue {
      return false
    }
    guard let conversationId = flat["conversationId"] ?? flat["conversation_id"],
          let messageId = flat["messageId"] ?? flat["message_id"],
          !conversationId.isEmpty, !messageId.isEmpty else {
      completionHandler(.failed)
      return true
    }
    let ok = NativeDeliveredAck.postDelivered(config: config, conversationId: conversationId, messageId: messageId)
    if !ok {
      PushNativeStore.enqueueAck(conversationId: conversationId, messageId: messageId)
    }
    PushEventStreamHandler.emit([
      "kind": "incoming_chat_message",
      "conversationId": conversationId,
      "messageId": messageId,
      "nativeAckSucceeded": ok,
    ])
    completionHandler(ok ? .newData : .failed)
    return true
  }

  private static func flattenUserInfo(_ userInfo: [AnyHashable: Any]) -> [String: String] {
    var out: [String: String] = [:]
    for (key, value) in userInfo {
      let k = String(describing: key)
      if let s = value as? String {
        out[k] = s
      } else if let nested = value as? [String: Any] {
        for (k2, v2) in nested {
          out[k2] = String(describing: v2)
        }
      } else {
        out[k] = String(describing: value)
      }
    }
    return out
  }
}
